onnxruntime/core/session/onnxruntime_c_api.cc

```diff
@@ -2243,6 +2248,9 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsSetCustomJoinThreadFn, _Inout_ OrtSes
 
 static constexpr OrtApiBase ort_api_base = {
     &OrtApis::GetApi,
+#ifdef ENABLE_TRAINING_ON_DEVICE
+    &OrtTrainingApis::GetTrainingApi,
+#endif
```

Comment:
This makes anyone calling GetVersionString without #define ENABLE_TRAINING_ON_DEVICE crash as they call the wrong function.

You also have an issue where anyone running training on the wrong build will also crash calling a nonexistent function.

You could have it so you pass a special version to GetApi() and add an inline function here to make it easy, like:

inline OrtTrainingApi* OrtGetTrainingApi() { return (OrtTrainingApi*)GetApiBase()->GetApi(ORT_API_VERSION+0x10000); }

This avoids adding a new API, and as long as GetApi() is documented properly internally it'll work well into the future.

Reply:
Could we get the training API from the existing API instead? 

That way the #ifdef on ENABLE_TRAINING_ON_DEVICE can be inside the implementation of that function and you can return an error for a build where it wasn't included.

i.e. usage would be something like Ort::GetApi().GetTrainingApi() if you went via the C++ API.

Reply:
Accepting @RyanUnderhill and @skottmckay's suggestion. Users can call `Ort::GetApi().GetTrainingApi()` to get the training API. Will return a `nullptr` if the build is not a training build.
cc @pranavsharma